### ISSUE TITLE
Design fee structure & price points for genesis parameter packages

### ISSUE THREAD

**Issue by arjunhassard** [nucypher/nucypher#3082]

**Comment by arjunhassard** [comment]
Three major factors influencing potential fee models (+ corresponding notes): 

1. _Genesis parameter packages, whitelist & right to modify_ (https://github.com/nucypher/nucypher-ts/issues/168#issuecomment-1530632826)
2. _Alice decides vs. Enrico decides  & impact of free-riding on network_(https://github.com/nucypher/nucypher/issues/3050)
   - Excessive free-riding on paid Strategies will consume various resources; most likely to impact Ursulas (and therefore paying adopters) are chain interactions (checking conditions, possibly Enrico authorizations) – i.e. hitting rate limits on Infura
      - Multicall bundling would help with this
    - Does payment-gating the checking of conditions obviate an Enrico allowlist? 
        - Note that Bob paying is not the same as Bob presenting a payment proof 
           - We currently have a 'global sponsorship' model 
           - If Enricos are 'blockchainy', then there's no need for an allowlist 
              - But preferable for Enrico to be passed a payment proof via a sponsor model (i.e. not blockhainy)
              - Receiving an API key for example would be out-of-band
              - Capturing a commission on a contract between data owner and encryptor is out of scope, for now
    - Future versions may offer cohort-contract pairs directly to adopters 
       - These should be cheaper than an API with arbitrary conditions (what we're launching with by dint of our private beta group requirements) 
3. _Future-proofing updatable conditions_ (https://github.com/nucypher/nucypher/issues/3053)
    - First we need to clarify the new roles in the CBD cryptological narrative
       - Alice is the cohort controller (initially) and the sponsor (in certain cases), but definitely NOT the data owner. 
        - Managing the authority to update conditions blurs the traditional ideas of Alice & Enrico 
           - Must avoid permanence of trust in Alice 
        - Currently the coordinator contract does not impose conditions on the DKG initiator 
    - One design involves Alice checksumming/hashing conditions and putting them on-chain. 
        - Then Enrico needs to encrypt with same conditions 
        - Anyone can pay for this service  
        - The question then becomes, who can change that incumbent hash? 
            - This is orthogonal to who paid for that condition update
            - Payment != authority
            - No point paying for unauthorized conditions 
         - The cohort initiator is not necessarily the condition updater
     - Fundamentally, are conditions decided by Alice or Enrico? 
        - Could a Cohort have a predefined & restricted set of possible condtions
        - Or is it arbitrary and Enrico is the authority and condition controller from the first moment they encrypt any payload for that Cohort? 
           - Enrico would have to put in designation of a hash via some interface 
     - When is it better from an adopter perspective to update conditions on the fly? 
     - When do the checks on condition state happen?
        - During request runtime
        - Every certain number of blocks 
     - Making every ciphertext backward-compatible may not be less work than having updatable conditions from genesis
     - 'Easiest' design is that conditions points to on-chain proxy contract 
        - Potentially leave it up to adopter to come up with an authority structure 

In order to build a fee structure proposal, we first need to answer: what exact authorities has the old (PRE) Alice evolved into? Which then leads to: which end-user actions are gated by or trigger payments?

Germane adopter feedback: 
https://github.com/nucypher/tdec/issues/127
https://github.com/nucypher/tdec/issues/126
https://github.com/nucypher/tdec/issues/128
https://github.com/nucypher/tdec/issues/4
https://github.com/nucypher/tdec/issues/13

**Comment by arjunhassard** [comment]
_When is it better from an adopter perspective to update conditions on the fly?_

Say a creator wishes to sell three flavors of access, power & trust with respect to their static asset: 

(1) No power, trustful, cheapest
Buyer can access asset (e.g. play the track), but decryption rights can be withdrawn at any time. This implies updatable conditions attached to ciphertext but no transfer of update authority. Buyer can transfer access by selling the NFT, but the new buyer will remain subject to these trust impositions (i.e. trusting the creator). 

(2) No power, trustless, mid-range
Buyer can access the asset persistently and trustlessly, and can sell/transfer those rights. This would involve the asset being (provably*) encrypted with immutable conditions – i.e. nobody can update conditions later. Alternatively, the creator can rent access for some pre-agreed time period – again, this would need to be enshrined in an immutable `ConditionSet`. 

(3) Powerful, trustless, most expensive
Buyer purchases both the decryption rights and the right to update associated conditions (slightly tautologous since the latter implies the former). This enables the buyer to effectively 'become' the creator – they can now modify conditions to anything they want. The obvious use case is 'buy-to-rent', where the buyer either updates the conditions to include an expiring/pay-to-extend NFT – note this would be equivalent to (1): trustful from the new buyer's perspective. Or, they encrypt the asset from scratch under (2) – impermanent but immutable rented access. 

(2) does not require updatable conditions 
(1) requires updatable conditions but is antithetical to the 'true ownership' value proposition, and undermines the value of using a trust-minimized access control layer from at least one side of the marketplace (buyer/renter side) 
(3) requires updatable conditions and a means to transfer authority**
however, since immutable conditions are likely to be a heavily-used option, in general and even within the context of this example, it may be easier to utilize the 'encrypt again locally' workaround. the exception to this is when the (effective) update authority needs to be transferred, but there's no viable way for the devices involved to download -> decrypt -> encrypt with new conditions

*not having updatable conditions at all is one way to 'prove' that all conditions encrypted under the CBD scheme are immutable 
**although self-revoking update authority is hard to prove. which means that each transfer of authority is actually expanding the set of buyers. impeding this is similar to the non-casual piracy problem (i.e. retaining an asset even after losing decryption rights), in that both are out of scope

**Comment by arjunhassard** [comment]
Despite complicating factors detailed in previous comments, the fundamental question has remained: What is the core service unit on which charges to CBD users are based, that best satisfies these objectives: https://github.com/nucypher/nucypher/issues/3057#issuecomment-1432433817

Here are three arguments for that core/base unit to be _time_, and for the API-level fee model (at least during the genesis/private beta era) to ignore other economic activity.   

**(1) Unpredictable and domain-specific activity can be better absorbed into contract-cohort pair offerings**

Access request frequency is highly variable by use case – they might occur 10x a year (crypto inheritance) or 10x a second (large multiplayer game). If requests were the base chargeable service unit, a given cost per request will either be unaffordable to higher throughput applications, and/or undercharge low throughput applications and potentially hurt post-subsidy staker sustainability. The distribution of request frequency across use case appears quite fat-tailed – there's currently no cluster of CBD adopters with very similar access behavior. 

Conversely, one can imagine a specialized set of contracts for crypto inheritance charging 500k GWEI (~$1) per access request, while equivalent custom contracts for accessing in-game assets at 5 GWEI per request (about $3.50 per player for 100 hours of game-time assuming 1 new request a second). In these examples, there's no straightforward way for a frugal inheritance beneficiary to circumvent the request fee, nor a particularly strong incentive given the value of the underlying data. The contract-captured proceeds can be distributed any way we like – see argument (3).   

Why charge/compensate for node availability then? Could the domain-specific contracts render base service units obsolete? This would be generally risky, because it would misalign a major ongoing cost of service with remuneration. But our roll-out/roadmap is relevant here too: we intend to move beyond a coordinator-contract enforced allowlist and enable permissionlessly generated public keys. Although this initialization will almost certainly involve a one-off (& node-reimbursing payment), it may be difficult to predict the future costs borne by the cohort. More importantly, if the nodes are not paid for some ‘base’ economic activity (being available, answering a decryption request, etc), the network would be vulnerable to any external actor who writes contracts that ping the network for service but exclude stakers from fees/commission collected. Note thanks to @jMyles for prompting this insight. 

In summary, cost-based pricing and a base service unit together form a necessary insurance against (a) griefing and (b) competitor free-riding, and a time-based fee structure seems compatible with a far wider set of use cases. 

**(2) Time-based fees can reduce the trust imposition on end-users in the adopter-Sponsor paradigm**  

It is highly likely that all private beta adopters will sponsor their end-user activity. This Sponsor role, and it’s implications for trust and power, may be inescapable over the medium-term too, given the onerousness of directly paying for infrastructure as a casual application user. 

(Adopters will also control the network parameters as the ‘Cohort Authority’, but this long-term problem is not addressable by the genesis fee model).   

While neither of these powers imply custody of the underlying data, like centralized threshold decryption, they do make data sharing – particularly with expectations of persistent decryption rights – vulnerable to denial of service, whether through malice or failure modes (e.g. the sponsoring entity goes into liquidation).   It would be helpful to insure against this potentially fatal disruption caused by a semi-trusted Sponsor suddenly discontinuing payment to Threshold nodes. 

An obvious solution is predicating all service provision on advanced payment. This mitigates (and disincentivizes) denial of service because if payment halts, there would be a public heads-up, and another entity or entities could step in as a replacement Sponsor (and spin up replica contracts if necessary). The end-users would retain access via persistent ConditionSets and the same public keys. 

It is easier to pay in advance based on time, rather than a 'spendable' unit like decryption request (or indeed, condition creation/modification). The heads-up duration could be depleted quickly by an attacker if the time buffer consisted of user actions/events. One could go further and add Verifiable Delay Functions here, but that’s probably overkill. 

**(3) Long-term sustainability of all nodes via fee rebates**

We have a de facto plan to distribute fees to all active nodes, perhaps weighted by both the cohorts they are servicing and relative stake size ('fee rebate'). This approach comes with various issues to resolve, but the positives include evening out a large variation in revenue from cohort-to-cohort – which can contribute to stake centralization and other non-ergodic trends. Regardless of where we land on this, a fee rebate model fits better with a time-based base service unit, because all nodes must aim for maximum availability, regardless of how many cohorts they are employed to. Any node that authorizes stake to CBD/TACo is effectively on permanent standby to replace any node that drop out of existing cohort, by choice (unbonding) or based on a modification by the Cohort Authority.

**Comment by jMyles** [comment]
All good points, @jdbertron.  And yeah, I think that's broad appetite on the team to consider projects which create substantial economy with comparatively low compute overhead (for projects like yours, potentially _extremely_ low) and to pass the savings on to them (or, to put it inversely, to charge the CPU burners their fare share).

We have arrived at *time* as the initially chargeable metric not because it is the closest fit for the situational crypto-economics (it obviously isn't) but because it's the most domain-unspecific metric that we can easily use to determine a base fee, and during this time, we can look at the actual functioning of the network to determine how to charge properly.  

Does that make sense?

**Comment by jMyles** [comment]
During a very intense call last night, we hammered out the following "yes and" with regard to time-as-base-metric logic:

* Paying for time - in advance - means at least two things:
** A cohort, bound by a DKG Ritual (an instantiation ceremony) to provide services specific to the payer (sponsor?), being available for the duration of the period paid
** Write-access to an Enrico-allowlist, which lawful nodes are bound to check and abide, for the duration of the period

The above is to be deployed on a ciphertext-forward-compatible, invite-only beta.

Here, "ciphertext-forward-compatible" means that, while we offer no vow regarding forward compatibility over control of the economics specifics of the Rituals, nor over the API, that subsequent to the conclusion of this beta, ciphertexts will remain valid for service on the network.

This comes with the proviso that nodes involved in ritual-bound cohorts ("covenants"?  "conventions"?) must retain their secret material, and be prepared to use it for the purposes of a yet-to-be-authored series of rituals which will enable the cohort to be refreshed with relief nodes, for the cohort to generate subsequent secret material via derivation in light of yet-to-be-determined material, and to reuse public transcripts and other public material as necessary to service these legacy ciphertexts.

It is possible that it may also favor, or even require, the use of a mass conversion machine, where offered ciphertexts are adapted (presumably via PRE) to availability as objects in a new regime, for which these higher-order rituals are a default.

In particular, this means that adopters choosing to handpick nodes, or to craft permissionless logic for nodes, are responsible for understanding and facilitating the custody of sufficient secret material among the nodes in question to perform the transition. 

Anything else?

**Comment by arjunhassard** [comment]
> Paying for time - in advance - means at least two things:
** A cohort, bound by a DKG Ritual (an instantiation ceremony) to provide services specific to the payer (sponsor?), being available for the duration of the period paid
** Write-access to an Enrico-allowlist, which lawful nodes are bound to check and abide, for the duration of the period
The above is to be deployed on a ciphertext-forward-compatible, invite-only beta.

This is a good summary of the genesis/beta fee model from an adopter perspective. 

> This comes with the proviso that nodes involved in ritual-bound cohorts ("covenants"? "conventions"?) must retain their secret material, and be prepared to use it for the purposes of a yet-to-be-authored series of rituals which will enable the cohort to be refreshed with relief nodes, for the cohort to generate subsequent secret material via derivation in light of yet-to-be-determined material, and to reuse public transcripts and other public material as necessary to service these legacy ciphertexts. It is possible that it may also favor, or even require, the use of a mass conversion machine, where offered ciphertexts are adapted (presumably via PRE) to availability as objects in a new regime, for which these higher-order rituals are a default. In particular, this means that adopters choosing to handpick nodes, or to craft permissionless logic for nodes, are responsible for understanding and facilitating the custody of sufficient secret material among the nodes in question to perform the transition.

Just to clarify this, as it may be quite hard to follow if you're not privy to all our internal discussions: 
- A small group of adopting projects will be invited to integrate the MVP version of CBD (v.7.0.0).
- These adopters will choose between two types of cohort – i. Stable and ii. Permissionless. See https://github.com/nucypher/nucypher-ts/issues/168#issuecomment-1530632826
- A unique public key (and unique cohort) will be generated for each adopter. App data will be encrypted to that public key. Adopters will have the exclusive right to modify _their_ cohort in the future (e.g. adding nodes). 
- Eventually, certain individual nodes will need to be replaced, for a variety of reasons. This replacement is achievable in multiple known ways, but current designs cost too much gas to make them worth implementing as is. Instead, we will continue optimizing this replacement process. In the meanwhile, cohorts are very unlikely to lose any meaningful number of nodes*. For example, a cohort of 30 may drop down to 28 active nodes after one year, but this will not affect the access control service (since the threshold will be 16). Once it is cheap enough to replace inactive nodes, this update will be rolled out and all cohorts can be made whole.
- Contrary to what is written above, adopters **are not responsible for facilitating the custody of secret material** among nodes. Nodes are responsible for handling the secret material associated with a given public key, and participating in the future rituals that replace inactive nodes. This can happen without adopters even knowing. But yes, it's good to understand what's going on. 
- The PRE-driven ciphertext 'mass conversion' idea may not be feasible, and (more likely) may not even be needed. Adopters don't need to think about this right now. 
- If/when we change some underlying cryptography in the future – like adding more stateful keystore stuff – cohorts can slowly migrate over to this new paradigm via the (now cheap) node replacement process mentioned above. 

*This is partially because we will set the unbonding period for CBD nodes as relatively lengthy, thus prolonging the time commitment of nodes to this network application.

**Comment by jMyles** [comment]
On a call today, we resolved:

1) That proper closure of this issue requires splitting it into multiple issues (as evinced by several of the comments above)
2) That the parameters are sufficiently understood for 7.0.0 that this issue can be moved to >=7.1.

**Comment by jMyles** [comment]
See also: #3197.
